pkg/tags/tag.go

```diff
@@ -0,0 +1,276 @@
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package tags
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/swarm/spancontext"
+	"github.com/opentracing/opentracing-go"
+)
+
+var (
+	errExists      = errors.New("already exists")
+	errNA          = errors.New("not available yet")
+	errNoETA       = errors.New("unable to calculate ETA")
+	errTagNotFound = errors.New("tag not found")
+)
+
+// State is the enum type for chunk states
+type State = uint32
+
+const (
+	StateSplit  State = iota // chunk has been processed by filehasher/swarm safe call
+	StateStored              // chunk stored locally
+	StateSeen                // chunk previously seen
+	StateSent                // chunk sent to neighbourhood
+	StateSynced              // proof is received; chunk removed from sync db; chunk is available everywhere
+)
+
+// Tag represents info on the status of new chunks
+type Tag struct {
+	Total  int64 // total chunks belonging to a tag
+	Split  int64 // number of chunks already processed by splitter for hashing
+	Seen   int64 // number of chunks already seen
+	Stored int64 // number of chunks already stored locally
+	Sent   int64 // number of chunks sent for push syncing
+	Synced int64 // number of chunks synced with proof
+
+	Uid       uint32        // a unique identifier for this tag
+	Anonymous bool          // indicates if the tag is anonymous (i.e. if only pull sync should be used)
+	Name      string        // a name tag for this tag
+	Address   swarm.Address // the associated swarm hash for this tag
+	StartedAt time.Time     // tag started to calculate ETA
+
+	// end-to-end tag tracing
+	ctx      context.Context  // tracing context
+	span     opentracing.Span // tracing root span
+	spanOnce sync.Once        // make sure we close root span only once
+}
+
+// NewTag creates a new tag, and returns it
+func NewTag(uid uint32, s string, total int64, anon bool) *Tag {
+	t := &Tag{
+		Uid:       uid,
+		Anonymous: anon,
+		Name:      s,
+		StartedAt: time.Now(),
+		Total:     total,
+	}
+
+	// context here is used only to store the root span `new.upload.tag` within Tag,
+	// we don't need any type of ctx Deadline or cancellation for this particular ctx
+	t.ctx, t.span = spancontext.StartSpan(context.Background(), "new.upload.tag")
```

Comment:
this probably needs to be broken down into a separate task to pull out the relevant wrappers from ethersphere/swarm into the tracing package no?

Reply:
I agree.